processing/src/main/java/io/druid/query/filter/BoundDimFilter.java

```diff
@@ -385,19 +385,20 @@ private void initLongData()
           if (hasLowerBound()) {
             final Long lowerLong = GuavaUtils.tryParseLong(lower);
             if (lowerLong == null) {
-              matchesAnything = false;
-              return;
+              // Unparseable values fall before all actual numbers, so all numbers will match the lower bound.
+              hasLowerLongBoundVolatile = false;
+            } else {
+              hasLowerLongBoundVolatile = true;
+              lowerLongBoundVolatile = lowerLong;
             }
-
-            hasLowerLongBoundVolatile = true;
-            lowerLongBoundVolatile = lowerLong;
           } else {
             hasLowerLongBoundVolatile = false;
```

Comment:
@gianm @leventov @himanshug 

Weighing in with my own $0.02, I tend to take the approach where if you are having to go through significant mental acrobatics to wrap your head around the happens-before relationship, then it's a good time to step back and look at exactly what you are trying to do.

In this case, I actually don't see a reason why you can't keep the "longs have been initialized" variable as volatile, and then after the first check, synchronize and then build the actual `DruidLongPredicate` inside the sychronization and cache that.  Then subsequent calls to `get()` just return that object.

From my read of that code, once the init happens once, it doesn't look like the state changes and if you pull all of that stuff up into the main method, then they are all method-local variables and it is really easy to understand their interplay and lifecycle.